### PR TITLE
Add documentation on broadcasting

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
The broadcasting feature lacks proper documentation. This is changed here, probably to be complemented by additional documentation or a demo.

**Description of the Change:**
This PR adds 
1. A brief section to the QNode quick start on how to use this technique, and pointers to the following documentation bits:
2. A "broadcasting" submenu to the QNode documentation with usage examples and a few gotchas and conventions.
3. A "broadcasting" submenu to the Operator documentation with implementation details and technical aspects.

**Benefits:**
Documentation coverage.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
N/A

### CODE REVIEW

File: pennylane/operation.py [hunk 1]
```diff
@@ -383,6 +383,98 @@ def circuit(angle):
     >>> circuit(a)
     -0.9999987318946099
 
+    .. details::
+        :title: Parameter broadcasting
```
Comment (josh146) [inline pennylane/operation.py:387]: :+1:

File: doc/introduction/circuits.rst [hunk 2]
```diff
@@ -58,18 +58,19 @@ Quantum functions are a restricted subset of Python functions, adhering to the f
 constraints:
 
 * The quantum function accepts classical inputs, and consists of
-  :doc:`quantum operators <operations>` or sequences of operators called :doc:`templates`,
-  using one instruction per line.
+  :doc:`quantum operators <operations>` or sequences of operators called :doc:`templates`.
```
Comment (dwierichs) [inline doc/introduction/circuits.rst:61]: Instructions like
```python
[qml.RX(x, wires=w) for w in dev.wires]
```
are allowed.

File: doc/introduction/circuits.rst [hunk 3]
```diff
@@ -58,18 +58,19 @@ Quantum functions are a restricted subset of Python functions, adhering to the f
...
 
 * The function can contain classical flow control structures such as ``for`` loops or ``if`` statements.
 
 * The quantum function must always return either a single or a tuple of
-  *measured observable values*, by applying a :doc:`measurement function <measurements>`
-  to a :ref:`qubit observable <intro_ref_ops_qobs>` or :ref:`continuous-value observable <intro_ref_ops_cvobs>`.
+  *measurement values*, by applying a :doc:`measurement function <measurements>`
+  to the qubit register. The most common example is to measure the expectation value of
+  a :ref:`qubit observable <intro_ref_ops_qobs>` or
+  :ref:`continuous-value observable <intro_ref_ops_cvobs>`.
```
Comment (dwierichs) [inline doc/introduction/circuits.rst:69]: Not only observable measurements are supported by PennyLane. In particular, `qml.var` includes post-processing of observable measurements.

File: doc/introduction/circuits.rst [hunk 4]
```diff
@@ -58,18 +58,19 @@ Quantum functions are a restricted subset of Python functions, adhering to the f
...
 
 .. note::
 
-    Quantum functions can only be evaluated on a device from within a QNode.
+    Quantum functions are evaluated on a device from within a QNode.
```
Comment (dwierichs) [inline doc/introduction/circuits.rst:73]: They typically are executed via QNodes, and it is the user-facing way to do it. But it is not the only one.

File: doc/introduction/circuits.rst [hunk 5]
```diff
@@ -114,7 +115,7 @@ The two most important device options are the ``wires`` and ``shots`` arguments.
 Wires
 *****
 
-The wires argument can either be an integer that defines the *number of wires*
+The wires argument can be an integer that defines the *number of wires*
```
Comment (dwierichs) [inline doc/introduction/circuits.rst:118]: This "either...or" construction was not closed.

File: doc/introduction/circuits.rst [hunk 6]
```diff
@@ -153,7 +154,8 @@ to estimate statistical quantities. On some supported simulator devices, ``shots
 measurement statistics *exactly*.
 
 Note that this argument can be temporarily overwritten when a QNode is called. For example, ``my_qnode(shots=3)``
-will temporarily evaluate ``my_qnode`` using three shots.
+will temporarily evaluate ``my_qnode`` using three shots. This is a feature of each QNode and it is not 
+necessary to manually implement the ``shots`` keyword argument of the quantum function.
```
Comment (dwierichs) [inline doc/introduction/circuits.rst:158]: For clarification.

File: doc/introduction/circuits.rst [hunk 7]
```diff
@@ -255,6 +257,41 @@ For example:
 
     result = circuit(0.543)
 
+Parameter Broadcasting in QNodes
+--------------------------------
+
+Some :class:`~.pennylane.QNode` circuits support the execution at multiple parameters
+simultaneously:
```
Comment (josh146) [inline doc/introduction/circuits.rst:264]: Maybe?
```suggestion
Depending on the quantum operations used, a :class:`~.pennylane.QNode` may support execution at multiple parameters simultaneously:
```

makes it less ambiguous wondering what causes the lack of support

File: doc/introduction/circuits.rst [hunk 8]
```diff
@@ -255,6 +257,41 @@ For example:
...
+machine learning).
+
+In addition to a more flexible execution syntax, broadcasting can yield performance boosts
+compared to the separate execution of the QNode for each parameter setting. Whether or not this is
+the case depends on quite a few details, but in particular for (at most) moderately sized circuits
+(:math:`\lesssim 20` wires) with a moderate number of parameters (:math:`\lesssim 200`) executed
+on a classical simulator, one can expect to benefit from broadcasting.
+See the :class:`~.pennylane.QNode` documentation for usage details.
+
+Many standard quantum operators support broadcasting, see the corresponding attribute
```
Comment (josh146) [inline doc/introduction/circuits.rst:282]:
```suggestion
Many standard quantum operators support broadcasting; see the corresponding attribute
```

File: doc/introduction/circuits.rst [hunk 9]
```diff
@@ -255,6 +257,41 @@ For example:
...
+simultaneously:
+
+>>> x = np.array([0.543, 1.234])
+>>> result = circuit(x)
+>>> result
+tensor([0.85616242, 0.33046511], requires_grad=True)
+
+This is called *parameter broadcasting* (as for, say, NumPy functions executed along an axis)
+or *parameter batching* (as in the application of a function to a *batch* of parameters in
+machine learning).
```
Comment (josh146) [inline doc/introduction/circuits.rst:273]:
```suggestion
Note that we are passing in a 1-dimensional array of parameters to the `circuit()`
QNode defined above, which takes a single parameter and returns a single expectation
value. As the input is now an array, the output is also an array, with each
element the expectation value of the corresponding input element.

This is called *parameter broadcasting* (as for, say, NumPy functions executed along an axis)
or *parameter batching* (as in the application of a function to a *batch* of parameters in
machine learning).
```

File: doc/introduction/circuits.rst [hunk 10]
```diff
@@ -255,6 +257,41 @@ For example:
...
+(:math:`\lesssim 20` wires) with a moderate number of parameters (:math:`\lesssim 200`) executed
+on a classical simulator, one can expect to benefit from broadcasting.
+See the :class:`~.pennylane.QNode` documentation for usage details.
+
+Many standard quantum operators support broadcasting, see the corresponding attribute
+:obj:`~.pennylane.ops.qubit.attributes.supports_broadcasting` for a list. The
+:class:`~.pennylane.operation.Operator` documentation contains implementation details
+and a guide to make custom operators compatible with broadcasting.
+Broadcasting can be used with any device, but will usually only yield performance upgrades for
+classical simulator devices like ``"default.qubit"`` that indicate that they support it:
```
Comment (josh146) [inline doc/introduction/circuits.rst:287]:
```suggestion
Broadcasting can be used with any device, but will usually only yield performance upgrades for
devices like ``"default.qubit"`` that indicate that they support it:
```

I have a feeling that even non-simulator devices will see speedups, partly due to avoiding slow python for loops.

File: pennylane/operation.py [hunk 11]
```diff
@@ -383,6 +383,105 @@ def circuit(angle):
...
+           above has ``ndim_params = (0,)`` for a single scalar argument.
+           An operator taking a matrix argument and a scalar would have ``ndim_params = (2, 0)``.
+           Note that ``ndim_params`` does *not require the size* of the axes.
+        #. Make the representations of the operator broadcasting-compatible. Typically, one or
+           multiple of the methods ``compute_matrix``, ``compute_eigvals`` and
+           ``compute_decomposition`` are defined by an operator, and these need to work with
+           the original input and output as well as with broadcasted inputs and outputs
+           that have an additional, leading axis. See below for an example.
+        #. Make sure that validation within the above representation methods and
+           ``__init__``--if it is overwritten by the operator class--allow
```
Comment (josh146) [inline pennylane/operation.py:409]:
```suggestion
           ``__init__``---if it is overwritten by the operator class---allow
```

File: pennylane/qnode.py [hunk 12]
```diff
@@ -148,15 +148,192 @@ class QNode:
...
+        We start with a simple example and briefly look at the scenarios in which broadcasting is
+        possible and useful. Finally we give rules and conventions regarding the usage of
+        broadcasting, together with some more complex examples.
+        Also see the :class:`~.pennylane.operation.Operator` documentation for implementation
+        details.
+
+        **Example**
+
+        Consider the ``circuit`` defined above. If we want to execute it at multiple values of
+        ``x``, we may pass those as a one-dimensional array to the QNode:
```
Comment (josh146) [inline pennylane/qnode.py:176]: @dwierichs if the reader has executed all codeblocks in order above, then `circuit` would not be a qnode :(

**Review by josh146 — APPROVED** [review]
Nice work @dwierichs! This is incredible helpful to have.

File: doc/introduction/circuits.rst (hunk 9 above)
Comment (dwierichs) [reply doc/introduction/circuits.rst:277]: Nice! :)

File: doc/introduction/circuits.rst (hunk 10 above)
Comment (dwierichs) [reply doc/introduction/circuits.rst]: Ah, yes, in the QNode documentation I mentioned this but here I do not explicitly mention it, so indeed it's good to not restrict the statement, I think :)

File: pennylane/qnode.py (hunk 12 above)
Comment (dwierichs) [reply pennylane/qnode.py:176]: Oh right, so I just want to write ``qnode`` instead of ``circuit``, correct? :)

File: pennylane/qnode.py [hunk 13]
```diff
@@ -148,15 +148,192 @@ class QNode:
...
+        *parameter broadcasting* or *parameter batching*.
+        We start with a simple example and briefly look at the scenarios in which broadcasting is
+        possible and useful. Finally we give rules and conventions regarding the usage of
+        broadcasting, together with some more complex examples.
+        Also see the :class:`~.pennylane.operation.Operator` documentation for implementation
+        details.
+
+        **Example**
+
+        Consider the ``circuit`` defined above. If we want to execute it at multiple values of
```
Comment (dwierichs) [inline pennylane/qnode.py:175]:
```suggestion
        Consider the ``qnode`` defined above. If we want to execute it at multiple values of

```

File: pennylane/qnode.py (hunk 12 above)
Comment (josh146) [reply pennylane/qnode.py:176]: Or maybe better to just define a qnode/circuit here.

File: pennylane/operation.py [hunk 14]
```diff
@@ -383,6 +383,105 @@ def circuit(angle):
...
+        cases separately.
+
+        .. code-block:: python
+
+            @staticmethod
+            def compute_decomposition(theta, wires):
+                '''Iterate over the last axis of theta, which is also the first axis
+                or the second axis without and with broadcasting, respectively.'''
+                decomp_ops = [qml.RX(x, wires=w) for x, w in zip(qml.math.T(theta), wires)]
+                return decomp_ops
```
Comment (albi3ro) [inline pennylane/operation.py:484]: Worth adding documentation on how `_check_batching` is supposed to work, answering such questions as:
* when does `_check_batching` need to be called? (any time the data is modified)
* what attributes does it set?
* What do those attributes look like if `_check_batching` is not called?
* What do those attributes look like if the Operator does not support batching?
* What does `_check_batching` do if the Operator does not support batching?

Comment (dwierichs) [reply pennylane/operation.py:484]: Great idea, thanks!

Comment (dwierichs) [reply pennylane/operation.py:487]: I added a paragraph on `_check_batching`. Happy to hear your thoughts on it!

**Comment by Jaybsoni** [comment]
@dwierichs and @albi3ro, 

Is this ready to be merged? Can I tag it with the `merge-ready` tag so we can ensure it gets in for the release today?

**Comment by josh146** [comment]
@Jaybsoni this is merge ready from my end.

**Comment by dwierichs** [comment]
Good to go from my end as well. Apologies, I planned to have this merged in earlier today already.
As it is a documentation PR, would be good to delay into next week if necessary due to high CI workload/merging queue.

**Pull request merged**